feat(ivy): i18n compiler support for element attributes

AndrewKushnir opened the pull request:
## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
```

## What is the new behavior?
This PR adds i18n support for element attributes at compile time.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

AndrewKushnir committed:
feat(ivy): i18n compiler support for element attributes

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview 5808c44 at https://pr26280-5808c44.ngbuilds.io/.

AndrewKushnir commented:
Hi @ocombe! Could you please have a look at this PR when you get a chance? I've seen your PR for Runtime part - I think we can converge our changes at some point. Thank you.

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview 1c7282d at https://pr26280-1c7282d.ngbuilds.io/.

alfaproject commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -495,6 +498,29 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
         this.creationInstruction(element.sourceSpan, R3.disableBindings);
       }
 
+      if (i18nInputs.length > 0) {
```
No need for this condition

AndrewKushnir replied:
Thanks for the note. I've updated the code to get rid of this check.

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview e5f17e4 at https://pr26280-e5f17e4.ngbuilds.io/.

ocombe commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -71,6 +74,21 @@ export function isI18NAttribute(name: string): boolean {
+  let acc = '';
+  const lastIdx = strings.length - 1;
+  for (let i = 0; i < lastIdx; i++) {
+    acc += `${strings[i]}${wrapI18nPlaceholder(i)}`;
```
I think that for closure you need to wrap variables with `${ ... }`, not with `�`., and then add the parameters wrapped with that symbol as the second argument of `goog.getMsg`.
See: https://github.com/angular/angular/blob/685712587db946295a75122580e1eeb1d7b80b2e/packages/core/src/render3/i18n.md#googgetmsg

For example:
```ts
/**
 * @desc Greeting.
 */
const MSG = goog.getMsg('Hello {$name}!', {
  name: '�0�'
});
```

AndrewKushnir replied:
Thanks for the review @ocombe, good catch! There is more logic that is coming with `i18nStart` instruction that would require changes to the way we generate `goog.getMsg` input, so I'd keep it as is for now and take care of this during `i18nStart` instruction implementation.
Please let me know if you're ok with that.

mhevery replied:
correct

ocombe replied:
sounds good!

ocombe reviewed (changes requested):
seems good, except for the variables of `goog.getMsg`

mhevery commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -482,7 +485,7 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
 
     const createSelfClosingInstruction = !hasStylingInstructions && !isNgContainer &&
-        element.children.length === 0 && element.outputs.length === 0;
+        element.children.length === 0 && element.outputs.length === 0 && i18nInputs.length === 0;
```
technically speaking we can still have self closing element. The `i18nAttribute(index, ...)` only does `index-` to get the element and so it will work just as well with self closing tag. But it is a bit weird so I will leave it up to you

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview 4187b5a at https://pr26280-4187b5a.ngbuilds.io/.

AndrewKushnir commented:
@mhevery @ocombe thanks for the review!

I've applied the changes according to the feedback:

- updated the `i18nAttribute` fn contract (to use single fn invocation)
- static labels (without interpolation) are also processed through `i18nAttribute` fn
- tests were also updated to reflect the latest version of code generation

Note: I kept the `goog.getMsg` invocation format the same for now - will update it accordingly while working on `i18nStart` instruction support.

Please have another look and let me know if you have additional feedback.

Thank you.

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview da3b1e9 at https://pr26280-da3b1e9.ngbuilds.io/.

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

mary-poppins commented:
You can preview 1752d8d at https://pr26280-1752d8d.ngbuilds.io/.

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -283,6 +283,22 @@ function appendI18nNode(
   return tNode;
 }
 
+export function i18nAttribute(index: number, attrName: string, attrValue: string): void {
```
This signature seems wrong. I think it needs to take an array per our discussion.

mhevery commented:
Clean up my comment, and it is LGTM from  my side.

AndrewKushnir committed:
fixup! feat(ivy): i18n compiler support for element attributes

AndrewKushnir commented:
Thanks @mhevery. I've updated the `i18nAttribute` function contract and rebased from the latest master.

mary-poppins commented:
You can preview cf8022c at https://pr26280-cf8022c.ngbuilds.io/.

mary-poppins commented:
You can preview e233718 at https://pr26280-e233718.ngbuilds.io/.
